File: assignment-1-basic-lexer/lexer.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

class Lexer {
    public static class DFA {
        HashSet<String> RESERVED_WORDS = new HashSet<>(
                Arrays.asList("TAGS", "BEGIN", "SEQUENCE", "INTEGER", "DATE", "END"));
        List<Integer> WHITESPACES = Arrays.asList(9, 10, 11, 12, 13, 32);
        List<Integer> SPECIALS = Arrays.asList((int) '"', (int) '|', (int) '(', (int) ')', (int) ',', (int) '{',
                (int) '}');
        List<Integer> NUMBERS = Arrays.asList(48, 49, 50, 51, 52, 53, 54, 55, 56, 57);

        List<Integer> UPPER_AZ = new ArrayList<>();
        List<Integer> LOWER_AZ = new ArrayList<>();
        HashMap<String, String> TOKEN_NAMES = new HashMap<>();

        enum State {
            WhiteSpace,

            TypeRef1,
            TypeRef2,

            Identifier1,
            Identifier2,

            Number1,
            Number2,

            Assign1,
            Assign2,
            Assign3,

            Range1,
            Range2,

            Special
        }

        State startState;
        Set<State> finalStates = new HashSet<>();
        Map<String, State> transitions = new HashMap<>();

        public DFA() {
            initAlphabets();
            initTokenNames();
            drawPartialDFA();
        }

        private void initAlphabets() {
            // a..z
            for (int i = 97; i <= 122; i++) {
                LOWER_AZ.add(i);
            }
            // A..Z
            for (int i = 65; i <= 90; i++) {
                UPPER_AZ.add(i);
            }
        }

        public String encodeTransition(State state, int characterInt) {
            return state.name() + " " + Integer.toString(characterInt);
        }

        public void initTokenNames() {
            // Whitespaces
            TOKEN_NAMES.put(Character.toString((char) 9), "HORIZONTAL TABULATION");
            TOKEN_NAMES.put(Character.toString((char) 10), "LINE FEED");
            TOKEN_NAMES.put(Character.toString((char) 11), "VERTICAL TABULATION");
            TOKEN_NAMES.put(Character.toString((char) 12), "FORM FEED");
            TOKEN_NAMES.put(Character.toString((char) 13), "CARRIAGE RETURN");
            TOKEN_NAMES.put(Character.toString((char) 32), "SPACE");

            // Specials
            TOKEN_NAMES.put(Character.toString('{'), "LCURLY");
            TOKEN_NAMES.put(Character.toString('}'), "RCURLY");
            TOKEN_NAMES.put(Character.toString(','), "COMMA");
            TOKEN_NAMES.put(Character.toString('('), "LPAREN");
            TOKEN_NAMES.put(Character.toString(')'), "RPAREN");
            TOKEN_NAMES.put(Character.toString('|'), "VLINE");
            TOKEN_NAMES.put(Character.toString('"'), "QUOTE");

            // Normal states
            TOKEN_NAMES.put(State.TypeRef1.name(), "TypeRef");
            TOKEN_NAMES.put(State.Identifier1.name(), "Identifier");
            TOKEN_NAMES.put(State.Number1.name(), "Number");
            TOKEN_NAMES.put(State.Number2.name(), "Number");
            TOKEN_NAMES.put(State.Assign3.name(), "ASSIGN");
            TOKEN_NAMES.put(State.Range2.name(), "Range_Separator");
        }

        public void drawPartialDFA() {
            // === 1. init states ===

            startState = State.WhiteSpace;

            finalStates.add(State.WhiteSpace);
            finalStates.add(State.TypeRef1);
            finalStates.add(State.Identifier1);
            finalStates.add(State.Number1);
            finalStates.add(State.Number2);
            finalStates.add(State.Assign3);
            finalStates.add(State.Range2);
            finalStates.add(State.Special);

            // === 2. init transitions ===

            drawMultiple(State.WhiteSpace, State.WhiteSpace, WHITESPACES);

            drawMultiple(State.WhiteSpace, State.TypeRef1, UPPER_AZ);
            drawMultiple(State.TypeRef1, State.TypeRef1, LOWER_AZ);
            drawMultiple(State.TypeRef1, State.TypeRef1, UPPER_AZ);
            drawMultiple(State.TypeRef1, State.TypeRef1, NUMBERS);
            drawMultiple(State.TypeRef1, State.WhiteSpace, WHITESPACES); // Final
            drawSingle(State.TypeRef1, State.TypeRef2, (int) '-');
            drawMultiple(State.TypeRef2, State.TypeRef1, LOWER_AZ);
            drawMultiple(State.TypeRef2, State.TypeRef1, UPPER_AZ);
            drawMultiple(State.TypeRef2, State.TypeRef1, NUMBERS);

            drawMultiple(State.WhiteSpace, State.Identifier1, LOWER_AZ);
            drawMultiple(State.Identifier1, State.Identifier1, LOWER_AZ);
            drawMultiple(State.Identifier1, State.Identifier1, UPPER_AZ);
            drawMultiple(State.Identifier1, State.Identifier1, NUMBERS);
            drawMultiple(State.Identifier1, State.WhiteSpace, WHITESPACES); // Final
            drawSingle(State.Identifier1, State.Identifier2, (int) '-');
            drawMultiple(State.Identifier2, State.Identifier1, LOWER_AZ);
            drawMultiple(State.Identifier2, State.Identifier1, UPPER_AZ);
            drawMultiple(State.Identifier2, State.Identifier1, NUMBERS);

            drawSingle(State.WhiteSpace, State.Number1, (int) '0');
            drawMultiple(State.Number1, State.WhiteSpace, WHITESPACES); // Final

            drawMultiple(State.WhiteSpace, State.Number2, NUMBERS.subList(1, NUMBERS.size())); // 1 .. 9
            drawMultiple(State.Number2, State.Number2, NUMBERS);
            drawMultiple(State.Number2, State.WhiteSpace, WHITESPACES); // Final

            drawSingle(State.WhiteSpace, State.Assign1, (int) ':');
            drawSingle(State.Assign1, State.Assign2, (int) ':');
            drawSingle(State.Assign2, State.Assign3, (int) '=');
            drawMultiple(State.Assign3, State.WhiteSpace, WHITESPACES); // Final

            drawSingle(State.WhiteSpace, State.Range1, (int) '.');
            drawSingle(State.Range1, State.Range2, (int) '.');
            drawMultiple(State.Range2, State.WhiteSpace, WHITESPACES); // Final

            drawMultiple(State.WhiteSpace, State.Special, SPECIALS);
            drawMultiple(State.Special, State.WhiteSpace, WHITESPACES); // Final
        }

        public void drawSingle(State start, State end, int characterInt) {
            transitions.put(encodeTransition(start, characterInt), end);
        }

        public void drawMultiple(State start, State end, List<Integer> characters) {
            for (int character : characters) {
                transitions.put(encodeTransition(start, character), end);
            }
        }

        public boolean accepts(String input) {
            State prevState = null;
            State currentState = startState;
            StringBuilder sb = new StringBuilder();

            for (int i = 0; i < input.length(); i++) {
                char character = input.charAt(i);

                // DEBUG
                // System.out.println(i + "-->" + (int) character + "-->" + character);
                // System.out.println("Before: " + currentState.name());

                // Escape right away if character is invalid => So our DFA model is not complete
                // but PARTIAL
                if (!transitions.containsKey(encodeTransition(currentState, character))) {
                    // DEBUG
                    // System.out.println(i + "-->" + (int) character + "-->" + character);
                    // System.out.println(currentState.name());
                    // System.out.println();

                    sb.append(character);
                    System.out.println("Token: [" + sb.toString() + "] --> " + "Invalid");

                    return false;
                }

                prevState = currentState;
                currentState = transitions.get(encodeTransition(currentState, character));

                // DEBUG
                // System.out.println("After: " + currentState.name());
                // System.out.println();

                if (sb.length() == 1 && TOKEN_NAMES.containsKey(sb.toString())) {
                    // Single character token
                    System.out.println("Token: [" + sb.toString() + "] --> [" + TOKEN_NAMES.get(sb.toString()) + "]");
                    sb = new StringBuilder(); // Reset for next token sequence
                } else if (sb.length() != 0 && prevState != currentState && finalStates.contains(prevState)) {
                    // Multi-character token
                    if (RESERVED_WORDS.contains(sb.toString())) {
                        System.out.println(
                                "Token: [" + sb.toString() + "] --> [" + "RESERVED" + " - " + sb.toString() + "]");
                    } else {
                        System.out
                                .println("Token: [" + sb.toString() + "] --> [" + TOKEN_NAMES.get(prevState.name())
                                        + "]");
                    }
                    sb = new StringBuilder(); // Reset for next token sequence
                }

                // Append current character to current token
                sb.append(character);
            }

            // DEBUG
            // System.out.println(currentState.name());
            // System.out.println();

            if (!finalStates.contains(currentState)) {
                System.out.println("Token: [" + sb.toString() + "] --> " + "Incomplete");
            }

            return finalStates.contains(currentState);
        }
    }

    public static String readInputFile(String path) throws IOException {
        Path filePath = Paths.get(path);

        try {
            String content = Files.readString(filePath);
            return content;
        } catch (IOException e) {
            e.printStackTrace();
            System.exit(1);
            return null;
        }
    }

    public static void main(String[] args) throws IOException {
        // 1. Read input
        String input = readInputFile(args[0]);

        // DEBUG
        // System.out.println(input);

        // 2. Create DFA
        DFA dfa = new DFA();
        System.out.println("Result: " + dfa.accepts(input));
    }
}

File: assignment-1-basic-lexer/test-cpp/lexer.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

class DFA {
  enum State {
    WhiteSpace,
    TypeRef1,
    TypeRef2,
    Identifier1,
    Identifier2,
    Number1,
    Number2,
    Assign1,
    Assign2,
    Assign3,
    Range1,
    Range2,
    Special,
  };

  unordered_set<string> RESERVED_WORDS = {"TAGS",    "BEGIN", "SEQUENCE",
                                          "INTEGER", "DATE",  "END"};
  vector<int> WHITESPACES = {9, 10, 11, 12, 13, 32};
  vector<int> SPECIALS = {(int)'"', (int)'|', (int)'(', (int)')',
                          (int)',', (int)'{', (int)'}'};
  vector<int> NUMBERS = {48, 49, 50, 51, 52, 53, 54, 55, 56, 57};

  vector<int> UPPER_AZ;
  vector<int> LOWER_AZ;
  unordered_map<string, string> TOKEN_NAMES;

private:
  State startState;
  unordered_set<State> finalStates;
  unordered_map<string, State> transitions;

public:
  DFA() {
    initAlphabets();
    initTokenNames();
    drawPartialDFA();
  }

  void initAlphabets() {
    // a..z
    for (int i = 97; i <= 122; i++) {
      LOWER_AZ.push_back(i);
    }
    // A..Z
    for (int i = 65; i <= 90; i++) {
      UPPER_AZ.push_back(i);
    }
  }

  string getStateName(State state) {
    switch (state) {
    case WhiteSpace:
      return "WhiteSpace";
    case TypeRef1:
      return "TypeRef1";
    case TypeRef2:
      return "TypeRef2";
    case Identifier1:
      return "Identifier1";
    case Identifier2:
      return "Identifier2";
    case Number1:
      return "Number1";
    case Number2:
      return "Number2";
    case Assign1:
      return "Assign1";
    case Assign2:
      return "Assign2";
    case Assign3:
      return "Assign3";
    case Range1:
      return "Range1";
    case Range2:
      return "Range2";
    case Special:
      return "Special";
    }
  }

  string encodeTransition(State state, int characterInt) {
    return getStateName(state) + " " + to_string(characterInt);
  }

  void initTokenNames() {
    TOKEN_NAMES[string(1, static_cast<char>(9))] = "HORIZONTAL TABULATION";
    TOKEN_NAMES[string(1, static_cast<char>(10))] = "LINE FEED";
    TOKEN_NAMES[string(1, static_cast<char>(11))] = "VERTICAL TABULATION";
    TOKEN_NAMES[string(1, static_cast<char>(12))] = "FORM FEED";
    TOKEN_NAMES[string(1, static_cast<char>(13))] = "CARRIAGE RETURN";
    TOKEN_NAMES[string(1, static_cast<char>(32))] = "SPACE";

    TOKEN_NAMES[string(1, '{')] = "LCURLY";
    TOKEN_NAMES[string(1, '}')] = "RCURLY";
    TOKEN_NAMES[string(1, ',')] = "COMMA";
    TOKEN_NAMES[string(1, '(')] = "LPAREN";
    TOKEN_NAMES[string(1, ')')] = "RPAREN";
    TOKEN_NAMES[string(1, '|')] = "VLINE";
    TOKEN_NAMES[string(1, '"')] = "QUOTE";

    TOKEN_NAMES[getStateName(TypeRef1)] = "TypeRef";
    TOKEN_NAMES[getStateName(Identifier1)] = "Identifier";
    TOKEN_NAMES[getStateName(Number1)] = "Number";
    TOKEN_NAMES[getStateName(Number2)] = "Number";
    TOKEN_NAMES[getStateName(Assign3)] = "ASSIGN";
    TOKEN_NAMES[getStateName(Range2)] = "Range_Separator";
  }

  void drawSingle(State start, State end, int characterInt) {
    transitions[encodeTransition(start, characterInt)] = end;
  }

  void drawMultiple(State start, State end, vector<int> characters) {
    for (int i = 0; i < characters.size(); i++) {
      int character = characters[i];
      transitions[encodeTransition(start, character)] = end;
    }
  }

  void drawPartialDFA() {
    startState = WhiteSpace;

    finalStates.insert(WhiteSpace);
    finalStates.insert(TypeRef1);
    finalStates.insert(Identifier1);
    finalStates.insert(Number1);
    finalStates.insert(Number2);
    finalStates.insert(Assign3);
    finalStates.insert(Range2);
    finalStates.insert(Special);

    drawMultiple(WhiteSpace, WhiteSpace, WHITESPACES);

    drawMultiple(WhiteSpace, TypeRef1, UPPER_AZ);
    drawMultiple(TypeRef1, TypeRef1, LOWER_AZ);
    drawMultiple(TypeRef1, TypeRef1, UPPER_AZ);
    drawMultiple(TypeRef1, TypeRef1, NUMBERS);
    drawMultiple(TypeRef1, WhiteSpace, WHITESPACES); // Final
    drawSingle(TypeRef1, TypeRef2, (int)'-');
    drawMultiple(TypeRef2, TypeRef1, LOWER_AZ);
    drawMultiple(TypeRef2, TypeRef1, UPPER_AZ);
    drawMultiple(TypeRef2, TypeRef1, NUMBERS);

    drawMultiple(WhiteSpace, Identifier1, LOWER_AZ);
    drawMultiple(Identifier1, Identifier1, LOWER_AZ);
    drawMultiple(Identifier1, Identifier1, UPPER_AZ);
    drawMultiple(Identifier1, Identifier1, NUMBERS);
    drawMultiple(Identifier1, WhiteSpace, WHITESPACES); // Final
    drawSingle(Identifier1, Identifier2, (int)'-');
    drawMultiple(Identifier2, Identifier1, LOWER_AZ);
    drawMultiple(Identifier2, Identifier1, UPPER_AZ);
    drawMultiple(Identifier2, Identifier1, NUMBERS);

    drawSingle(WhiteSpace, Number1, (int)'0');
    drawMultiple(Number1, WhiteSpace, WHITESPACES); // Final

    vector<int> subvector(NUMBERS.begin() + 1, NUMBERS.end());
    drawMultiple(WhiteSpace, Number2, subvector);
    drawMultiple(Number2, Number2, NUMBERS);
    drawMultiple(Number2, WhiteSpace, WHITESPACES); // Final

    drawSingle(WhiteSpace, Assign1, (int)':');
    drawSingle(Assign1, Assign2, (int)':');
    drawSingle(Assign2, Assign3, (int)'=');
    drawMultiple(Assign3, WhiteSpace, WHITESPACES); // Final

    drawSingle(WhiteSpace, Range1, (int)'.');
    drawSingle(Range1, Range2, (int)'.');
    drawMultiple(Range2, WhiteSpace, WHITESPACES); // Final

    drawMultiple(WhiteSpace, Special, SPECIALS);
    drawMultiple(Special, WhiteSpace, WHITESPACES); // Final
  }

  bool accepts(string input) {
    State prevState = startState;
    State currentState = startState;
    ostringstream oss; // string builder
    string currentString = "";

    for (int i = 0; i < input.length(); i++) {
      char character = input.at(i);

      if (transitions.find(encodeTransition(currentState, character)) ==
          transitions.end()) {

        oss << character;
        currentString = oss.str();

        cout << "Token: [" << currentString << "] --> " << "Invalid" << endl;
        return false;
      }

      prevState = currentState;
      currentState = transitions[encodeTransition(currentState, character)];

      if (currentString.size() == 1 &&
          TOKEN_NAMES.find(currentString) != TOKEN_NAMES.end()) {

        cout << "Token: [" << currentString << "] --> ["
             << TOKEN_NAMES[currentString] << "]" << endl;

        oss.str(""); // Clear the contents
        oss.clear(); // Reset the error state
      }

      else if (currentString.size() != 0 && prevState != currentState &&
               finalStates.find(prevState) != finalStates.end()) {

        if (RESERVED_WORDS.find(currentString) != RESERVED_WORDS.end()) {
          cout << "Token: [" << currentString << "] --> [" << "RESERVED"
               << " - " << currentString << "]" << endl;
        } else {
          cout << "Token: [" << currentString << "] --> ["
               << TOKEN_NAMES[getStateName(prevState)] << "]" << endl;
        }

        oss.str(""); // Clear the contents
        oss.clear(); // Reset the error state
      }

      oss << character;
      currentString = oss.str();
    }

    if (finalStates.find(currentState) == finalStates.end()) {
      cout << "Token: [" << currentString << "] --> " << "Incomplete" << endl;
    }

    return finalStates.find(currentState) != finalStates.end();
  }
};

string readInputFile(string path) {
  ifstream file(path); // Open the file
  if (!file) {         // Check if the file opened successfully
    throw runtime_error("Could not open the file: " + path);
  }

  ostringstream oss;   // Create a string stream
  oss << file.rdbuf(); // Read the file content into the stream

  return oss.str(); // Convert the stream to a string
}

int main(int argc, char *argv[]) {
  string input = readInputFile(argv[1]);

  DFA dfa = DFA();
  bool result = dfa.accepts(input);
  cout << "Result: " << result << endl;

  return 0;
}
